### PR TITLE
Feature: add ExprBase and add first tests with coverage

### LINKED ISSUE #69: Implement `gil::ast::ExprBase` base class

### PR DESCRIPTION
**Pull request by LindonAliu**
fix #69

### CODE REVIEW

File: test/test_ASTNode.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+TEST_F(ASTNodeTest, StmtBaseConstructor)
+{
+    StmtBase stmt(NodeKind::StmtBaseFirstKind, loc, nullptr);
+    ASSERT_EQ(stmt.getKind(), NodeKind::StmtBaseFirstKind);
+    ASSERT_TRUE(StmtBase::classof(&stmt));
+}
+
+TEST_F(ASTNodeTest, ExprBaseConstructor)
+{
+    ExprBase expr(NodeKind::ExprBaseFirstKind, loc, nullptr);
```
Comment (Snowy1803) [inline test/test_ASTNode.cpp:35]: that shouldn't be legal

File: include/AST/ASTNode.hpp [hunk 2]
```diff
@@ -37,27 +36,22 @@ class ASTNode {
     NodeKind _nodeKind;
 
 public:
-    ASTNode(
-        NodeKind kind, SourceLocation nodeLocation, ASTNode *parent = nullptr
-    )
-        : _parent(parent), _nodeKind(kind), _nodeLocation(nodeLocation)
+    ASTNode(NodeKind kind, SourceLocation nodeLocation, ASTNode *parent)
```
Comment (Snowy1803) [inline include/AST/ASTNode.hpp:39]: should be protected as well

File: test/test_ASTNode.cpp [hunk 3]
```diff
@@ -10,53 +10,77 @@ class ASTNodeTest : public ::testing::Test {
...
+    TestExprBase(NodeKind kind, glu::SourceLocation loc, ASTNode *parent)
+        : ExprBase(kind, loc, parent)
+    {
+    }
+};
+
 TEST_F(ASTNodeTest, ASTNodeConstructor)
 {
-    ASTNode node(NodeKind::DeclBaseFirstKind, loc);
+    ASTNode node(NodeKind::DeclBaseFirstKind, loc, nullptr);
```
Comment (Snowy1803) [inline test/test_ASTNode.cpp:39]: shouldn't be possible — doesn't make sense

File: test/test_ASTNode.cpp [hunk 4]
```diff
@@ -10,53 +10,77 @@ class ASTNodeTest : public ::testing::Test {
...
 
 TEST_F(ASTNodeTest, DeclBaseClassof)
 {
-    DeclBase decl(NodeKind::DeclBaseFirstKind, loc);
+    TestDeclBase decl(NodeKind::FunctionDeclKind, loc, nullptr);
     ASSERT_TRUE(DeclBase::classof(&decl));
-    ASTNode node(NodeKind::StmtBaseFirstKind, loc);
-    ASSERT_FALSE(DeclBase::classof(&node));
+    ASSERT_FALSE(StmtBase::classof(&decl));
+    ASSERT_FALSE(ExprBase::classof(&decl));
```
Comment (Snowy1803) [inline test/test_ASTNode.cpp:69]: use `llvm::isa<ExprBase>(decl)` instead of using classof directly

Comment (Snowy1803) [reply test/test_ASTNode.cpp:80]: you should really never use classof directly

Comment (LindonAliu) [reply test/test_ASTNode.cpp:80]: i've forgot sryyyyyyyyyyy

**Review by Snowy1803 — DISMISSED** [review]

**Review by yunsecode — DISMISSED** [review]

**Review by jeremyelalouf — APPROVED** [review]

**Review by yunsecode — APPROVED** [review]

**Pull request merged**